stats.go

```diff
@@ -57,81 +57,244 @@ func (c *capabilities) Tagging() bool {
 }
 
 type counter struct {
-	prev        int64
-	curr        int64
-	cachedCount CachedCount
-}
-
-func newCounter(cachedCount CachedCount) *counter {
-	return &counter{cachedCount: cachedCount}
+	prev           int64
+	curr           int64
+	cachedCount    CachedCount
+	lastUpdateUnix int64
+	expired        uint32
+	scope          *scope
+	name           string
+}
+
+func newCounter(cachedCount CachedCount, name string, scope *scope) *counter {
+	return &counter{
+		cachedCount:    cachedCount,
+		scope:          scope,
+		name:           name,
+		lastUpdateUnix: globalNow().Unix(),
+	}
 }
 
 func (c *counter) Inc(v int64) {
 	atomic.AddInt64(&c.curr, v)
+
+	if atomic.LoadUint32(&c.expired) == 1 && c.scope != nil {
+		// Counter has expired, but direct ref was held and counter
+		// was incremented, therefore insert back into scope
+		scope := c.scope.getCurrentScope()
+
+		scope.cm.Lock()
+		if counters, ok := scope.counters[c.name]; !ok {
+			// No counters with the same name were created since expiry
+			scope.counters[c.name] = []*counter{c}
+		} else {
+			exists := false
+			for _, counter := range counters {
```

Comment:
can avoid this check if instead, you did (after the if expired check):

```go
if !atomic.CompareAndSwapUint32(&c.expired, 1, 0) {
  // lost the race, another goroutine already added this counter back to the scope
  return
}

// now assume that we're not racing against ourself to add counter to `scope.counters[c.name]`

Reply:
Yeah good call. I can also use:
if atomic.CompareAndSwapUint32(&c.expired, 1, 0) instead of if atomic.LoadUint32(&c.expired) == 1 in the original check, but CompareAndSwapUint32 is less performant than LoadUint32 :(